### PR TITLE
[shardformer] chatglm support flash attention and jit operator

### PR DESCRIPTION
**Pull request by flybird1111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4184 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

chatglm support flash attention and jit operator

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/modeling/chatglm.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import torch
+import torch.nn.functional as F
+
+
+def get_flash_core_attention_forward():
+
+    from colossalai.kernel.cuda_native.flash_attention import AttnMaskType, ColoAttention
+    from tests.kit.model_zoo.transformers.chatglm2_6b.modeling_chatglm import CoreAttention
```
Comment (ver217) [inline colossalai/shardformer/modeling/chatglm.py:8]: You should not import any component of `tests` in colossalai.

File: colossalai/shardformer/modeling/chatglm.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            context_layer = context_layer.permute(1, 0, -1).contiguous()
+
+        return context_layer
+
+    return forward
+
+
+def get_jit_fused_glm_block_forward():
+
+    from tests.kit.model_zoo.transformers.chatglm2_6b.modeling_chatglm import GLMBlock
```
Comment (ver217) [inline colossalai/shardformer/modeling/chatglm.py:64]: Same problem

File: colossalai/shardformer/modeling/chatglm.py (hunk 1 above)
Comment (flybird1111) [reply colossalai/shardformer/modeling/chatglm.py:8]: > You should not import any component of `tests` in colossalai.


The chatglm model is not defined in the transformers, but is defined in the pretrain weights package of "THUDM/chatglm2-6b", Therefore it cannot be imported directly. we have placed the chatglm model file under the "test.kit". The modules are all imported from "tests.kit."

Comment (ver217) [reply colossalai/shardformer/modeling/chatglm.py:8]: `tests` is not a valid package. If you need, you should define it in colossalai and import it in `tests.kit`

Comment (ver217) [reply colossalai/shardformer/modeling/chatglm.py:8]: You should never import anything defined in `tests` unless you're writing unit tests.

Comment (flybird1111) [reply colossalai/shardformer/modeling/chatglm.py]: > You should never import anything defined in `tests` unless you're writing unit tests.

Updated.

File: colossalai/shardformer/modeling/chatglm.py (hunk 2 above)
Comment (flybird1111) [reply colossalai/shardformer/modeling/chatglm.py]: > Same problem

Updated.

**Review by ver217 — APPROVED** [review]

**Comment by flybird1111** [comment]
image
All tests have passed.

**Pull request merged**